libraries/chain/controller.cpp

```diff
@@ -4007,8 +4020,13 @@ void controller::get_finalizer_state( finalizer_state& fs ) const {
 }
 
 // called from net threads
-void controller::notify_hs_message( const uint32_t connection_id, const hs_message& msg ) {
-   my->pacemaker->on_hs_msg(connection_id, msg);
+void controller::notify_vote_message( const hs_vote_message& vote ) {
+   auto bsp = fetch_block_state_by_id(vote.proposal_id);
+   if( bsp ) {
+      my->finality_control.aggregate_vote(bsp, vote);
+   } else {
```

Comment:
Instead of adding `fetch_block_state_by_id` and `apply_block_data_new`, I suggest you add  a new method in `block_data_t` named:

`void aggregate_vote(const hs_vote_message& vote)` which would do nothing for the `dpos` case, and would fetch the block_state and apply the vote in the `if` case.

Reply:
Is `block_data_t` a general class and its main purpose to service fork_db APIs? `aggregate_vote` seems too specific for `block_data_t`.

Reply:
Block data and the other structs with variants should contain, as much as possible, any code that does something different for `dpos` and `if`, or accesses the block_state_* structs, so that the rest of controller doesn't see these differences.

It is perfectly fine to add new members like `aggregate_vote` which behave differently for `dpos` and `if`. 

The `apply` functions were mostly present to make controller work in the `dpos` mode while changes were not completed, ideally they should go away.